pkg/shp/cmd/build/run.go

```diff
@@ -71,6 +81,49 @@ func (c *RunCommand) Run(params *params.Params, ioStreams *genericclioptions.IOS
 		return err
 	}
 
-	fmt.Fprintf(ioStreams.Out, "BuildRun created %q\n", buildRun.Name)
+	if !c.Follow {
+		fmt.Fprintf(ioStreams.Out, "BuildRun created %q\n", buildRun.Name)
+		return nil
+	}
+
+	err := wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
```

Comment:
Should we create this mechanism based on a watch? And given the events modifying the pod state we can start and stop streaming logs?

Reply:
Interesting idea.  My first thought was I did not want to stop waiting after a predetermined time (perhaps making if configurable, though I did not do that at the moment), and so the poll lent itself better to enforcing a timeout.

But a watch is less impactful on the api server (thought this polling should not be onerous).

That said, I'm curious how the watch would play out.

I'll experiment with it some and report back.

thanks

Reply:
yeah @otaviof so replacing the poll immediate call with a build run watch to confirm the task run ref is set was pretty straight forward ... I just posted a separate commit for this.

Moving beyond those changes, and going into pod state tracking resulted again in a lot of duplication of code I'm getting for free from tkn.

Allow me to post some tkn code links to help illustrate my point, and confirm if the tkn logic I point you to is in fact what you were thinking about.

So my code calls https://github.com/tektoncd/cli/blob/0c7916aaf422ecc9208d3cb661f816a52b9190f1/pkg/log/reader.go#L83

So from there we go:
- to https://github.com/tektoncd/cli/blob/0c7916aaf422ecc9208d3cb661f816a52b9190f1/pkg/log/task_reader.go#L47-L59 ... notice the `isDone` check and toggle between live and available logs
- https://github.com/tektoncd/cli/blob/0c7916aaf422ecc9208d3cb661f816a52b9190f1/pkg/log/task_reader.go#L79-L86 which handles the translation to a pod and establishes the channels needed for live tracking
- then all the golang channel manipulation magic is occurring in https://github.com/tektoncd/cli/blob/0c7916aaf422ecc9208d3cb661f816a52b9190f1/pkg/log/task_reader.go#L163-L209

Reply:
Thanks for sharing the interesting bits from Tekton, @gabemontero . I think it gives a good head start, and specially the Printer module looks good 👍🏼